File: Week 7/java/1.StringBasics.java
/*
 * WEEK 7 - STRINGS
 * Topic: String Basics in Java
 *
 * KEY CONCEPTS:
 * - String in Java is IMMUTABLE — every modification creates a new object.
 * - StringBuilder is MUTABLE — efficient for repeated modifications.
 * - String pool: Java caches string literals for memory efficiency.
 *
 * IMPORTANT METHODS:
 * length(), charAt(), substring(), indexOf(), contains(), replace(),
 * toLowerCase(), toUpperCase(), trim(), split(), toCharArray()
 *
 * Time Complexity of common operations:
 * - charAt(i):      O(1)
 * - substring(l,r): O(r-l)  — creates a new string
 * - indexOf:        O(n*m) naive, O(n) with KMP
 * - concat (+):     O(n) — creates new string each time!
 * - StringBuilder append: O(1) amortized
 *
 * GOLDEN RULE: Never concatenate strings inside a loop with +.
 *              Use StringBuilder instead → O(n) vs O(n²)
 */

public class StringBasics {

    public static void main(String[] args) {
        // --- Basic Operations ---
        String s = "Hello, World!";
        System.out.println("String: " + s);
        System.out.println("Length: " + s.length());
        System.out.println("charAt(7): " + s.charAt(7));
        System.out.println("substring(7, 12): " + s.substring(7, 12));
        System.out.println("indexOf('o'): " + s.indexOf('o'));
        System.out.println("lastIndexOf('o'): " + s.lastIndexOf('o'));
        System.out.println("contains(\"World\"): " + s.contains("World"));
        System.out.println("toLowerCase: " + s.toLowerCase());
        System.out.println("toUpperCase: " + s.toUpperCase());
        System.out.println("replace('l', 'r'): " + s.replace('l', 'r'));
        System.out.println("trim: \"  hello  \".trim() = \"" + "  hello  ".trim() + "\"");

        // --- Comparison ---
        String a = "apple";
        String b = "apple";
        String c = new String("apple");
        System.out.println("\n\"apple\" == \"apple\": " + (a == b));       // true (same pool ref)
        System.out.println("\"apple\" == new String(\"apple\"): " + (a == c)); // false! different object
        System.out.println("\"apple\".equals(new String(\"apple\")): " + a.equals(c)); // true (content)
        System.out.println("compareTo: " + "apple".compareTo("banana")); // negative (a < b)

        // --- String to char array and back ---
        char[] chars = s.toCharArray();
        System.out.println("\ntoCharArray: " + chars.length + " chars");
        System.out.println("new String(chars): " + new String(chars));

        // --- Split ---
        String csv = "one,two,three,four";
        String[] parts = csv.split(",");
        System.out.println("\nSplit by ',': ");
        for (String p : parts) System.out.println("  " + p);

        // --- StringBuilder: efficient string building ---
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= 5; i++) {
            sb.append(i);
            if (i < 5) sb.append("-");
        }
        System.out.println("\nBuilt with StringBuilder: " + sb.toString());

        // Reverse with StringBuilder
        StringBuilder sb2 = new StringBuilder("hello");
        System.out.println("Reversed: " + sb2.reverse().toString());

        // --- BAD vs GOOD string concatenation in loops ---
        // BAD:  O(n²) — creates n new strings
        // String result = "";
        // for (int i = 0; i < n; i++) result += chars[i];

        // GOOD: O(n) — amortized O(1) per append
        StringBuilder good = new StringBuilder();
        for (char ch : chars) good.append(ch);
        System.out.println("\nRebuilt string: " + good.toString());
    }
}

File: Week 7/python/1.string_basics.py
"""
WEEK 7 - PYTHON DSA
Topic: String Basics
File: 1.string_basics.py

CONCEPT:
    Python strings are immutable sequences of Unicode code points.
    Every "modification" produces a new string. For repeated
    construction, build with a list and ''.join() or use io.StringIO.

KEY POINTS:
    - len(s), s[i], s[a:b], s.find(sub), 'sub' in s
    - .lower(), .upper(), .replace(), .strip(), .split()
    - Concatenation with `+` in a loop is O(n^2) like Java — use .join.
    - Strings are immutable; convert to list for in-place edits.

ALGORITHM / APPROACH:
    Common operations in O(1) or O(n):
        s[i]      O(1)
        s[a:b]    O(b-a)  copies into a new string
        s.find    O(n*m)  naive; CPython uses a fast Boyer-Moore variant
        ''.join() O(n)    when given a sequence of strings

PYTHON-SPECIFIC NOTES:
    - String "comparison" with == checks content; `is` checks identity.
    - f-strings (PEP 498) are the modern formatting tool.
    - .split() with no args splits on any whitespace and trims empties.

DRY RUN:
    s = "Hello, World!"
    len(s) = 13
    s[7] = 'W'
    s[7:12] = 'World'
    s.find('o') = 4
    s.lower() = 'hello, world!'
    s.replace('l','r') = 'Herro, Worrd!'

COMPLEXITY:
    See per-operation table above; most are linear in their argument size.
"""


def main() -> None:
    s = "Hello, World!"
    print(f"String: {s}")
    print(f"len: {len(s)}")
    print(f"s[7]: {s[7]}")
    print(f"s[7:12]: {s[7:12]}")
    print(f"s.find('o'): {s.find('o')}")
    print(f"s.rfind('o'): {s.rfind('o')}")
    print(f"'World' in s: {'World' in s}")
    print(f"s.lower(): {s.lower()}")
    print(f"s.upper(): {s.upper()}")
    print(f"s.replace('l','r'): {s.replace('l', 'r')}")
    print(f'  "  hello  ".strip() = "{"  hello  ".strip()}"')

    # --- Comparison ---
    a = "apple"
    b = "apple"
    c = str(["a", "p", "p", "l", "e"])  # different content
    print(f'\n"apple" == "apple": {a == b}')
    print(f'"apple" is "apple" (interned): {a is b}')
    print(f'a == c: {a == c}')

    # Lexicographic compare
    print(f'"apple" < "banana": {"apple" < "banana"}')

    # --- String <-> list ---
    chars = list(s)
    print(f"\nlist(s): {chars}")
    print(f"''.join(chars): {''.join(chars)}")

    # --- Split ---
    csv = "one,two,three,four"
    parts = csv.split(",")
    print("\nSplit by ',':")
    for p in parts:
        print(f"  {p}")

    # --- Build with join (preferred over += in a loop) ---
    pieces: list[str] = []
    for i in range(1, 6):
        pieces.append(str(i))
        if i < 5:
            pieces.append("-")
    print(f"\nBuilt with join: {''.join(pieces)}")

    # Reverse
    print(f"Reversed: {s[::-1]}")
    print(f"Reversed via reversed(): {''.join(reversed(s))}")

    # BAD: O(n^2) loop concat — shown but commented out
    # result = ""
    # for ch in chars: result += ch
    # GOOD: O(n) join
    good = "".join(chars)
    print(f"\nRebuilt string: {good}")


if __name__ == "__main__":
    main()


"""
NOTES — Python vs Java:
    - No StringBuilder; use a list + ''.join() for O(n) building.
    - Strings are sequences, so slicing (s[a:b], s[::-1]) is built in.
    - Unicode is native — len(s) counts code points, not bytes.
    - 'in' operator (`"x" in s`) replaces s.contains("x").
"""
